`upsample_nearest3d` and `upsample_trilinear3d`: port to structured

bdhirsh opened the pull request:
Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52065 port upsample_nearest3d and upsample_trilinear3d to structured**
* #52012 port upsample_bilinear2d and upsample_bicubic2d to structured
* #51917 port upsample_linear1d to structured
* #51916 remove unnecessary/dead code in upsample_nearest1d cuda kernel

Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

bdhirsh committed:
port upsample_nearest3d and upsample_trilinear3d to structured

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

ezyang commented:
> Also, a quick "gotcha" I hit when migrating to structured: If you forget to add structured: True in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway

This should be easy enough to add an explicit error check for.

ezyang commented:
@bdhirsh can you file a fresh issue about what to do about structured and quantized? That way we won't lose the discussion about it when this PR lands

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh commented:
> This should be easy enough to add an explicit error check for.

Hmm, where would that check go? I'm talking about the situation where, during porting, you write your kernels to use the `TORCH_IMPL_FUNC` macro et al., but forget to include the corresponding `structured: True` for your `out` ops in `native_functions.yaml`. Codegen can't complain earlier about the fact that your op isn't structured according to native_functions, since it doesn't see the kernel implementation. Unless I'm missing something. Also, I think this is fine to add as a "gotcha" on the porting guide if we can't directly add a better error message.

> can you file a fresh issue about what to do about structured and quantized? That way we won't lose the discussion about it when this PR lands

yep!

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

ezyang commented:
> Hmm, where would that check go? I'm talking about the situation where, during porting, you write your kernels to use the TORCH_IMPL_FUNC macro et al., but forget to include the corresponding structured: True for your out op in native_functions.yaml. Codegen can't complain earlier about the fact that your op isn't structured according to native_functions, because for all it knows, your op might not actually be structured (since it doesn't see the kernel implementation). Unless I'm missing something. Also, I think this is fine to add as a "gotcha" on the porting guide if we can't directly add a better error message.

Oh, well, if you forget to put structured anywhere, then I agree there's no way to test. But a `structured_delegate` should always point to a `structured: True` kernel, and I was a bit surprised that this wasn't tested (maybe you are saying that in the erroneous version, there wasn't `structured_delegate` either)

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh commented:
> Oh, well, if you forget to put structured anywhere, then I agree there's no way to test. But a `structured_delegate` should always point to a `structured: True` kernel, and I was a bit surprised that this wasn't tested (maybe you are saying that in the erroneous version, there wasn't `structured_delegate` either)

Oh yep, that's right. It looks like there isn't currently any error checking for that case. I can add one in, but it'll require making assertions across multiple native functions, which I don't think we currently do anywhere.

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

bdhirsh committed:
Update on "port upsample_nearest3d and upsample_trilinear3d to structured"

Porting `upsample_nearest3d` and `upsample_trilinear3d` to be structured kernels.

One annoying thing: since the `vec` overload of `upsample_nearest3d` has a `QuantizedCPU` kernel, we can't make it a `DefaultBackend` op without first updating the structured kernel codegen to handle `QuantizedCPU`. I still ended up updating that overload to redispatch and hit the structured version, since that still gets rid of a lot of code duplication (the shape checking), just not all of it.

A quick look showed 84 ops that have `QuantizedCPU` kernels. So it probably isn't a blocker to push off trying to get them to work with structured kernels.

A quick thought on updating structured kernels to deal with QuantizedCPU (and other keys). It's not as simple as adding another path to create the `out` Quantized tensor, because the arguments used to create the tensor are different than CPU/CUDA (there are extra quantized fields pulled off the tensor). Two quick options that come to mind:
- Update the signature of `MetaBase:set_output()` and `{op_name}:meta()` to be generic enough to handle all the cases: just take in a tensor, and leave it up to the op writers to pull the right fields off the tensor to generate the output. They would also need to plumb that tensor into `set_output`. **Downsides:** It shifts the onus of more logic from the codegen to kernel writers, since the interface is more generic. It would also require fixing up the existing meta kernels, so we'd have to do it sooner rather than later. **Upside:** We could probably maintain a single meta function per op, instead of making kernel writers implement a separate meta function for QuantizedCPU (and potential other keys later)
- Make an alternate base class `QuantizedMetaBase:set_output()` that has the corresponding signature required to set up out quantized tensors. And alternate classes `{op_name}_quantized:meta()` for each op with the correct `meta()` signature that all ops with a quantized kernel will need to implement. **Downsides:** Mildly more confusing class hierarchy, might need to expand it further for other keys. Some ops will need to implement two meta functions, although that's not really a huge deal since they can factor out the shared logic. **Upside:** Won't require a refactor of existing structured kernels.

I kind of like option 2 more, but I think they're both viable options. It also wouldn't require implementing immediately.

Also, a quick "gotcha" I hit when migrating to structured: If you forget to add `structured: True` in the right spot in native_functions.yaml, you get compiler errors about a missing class. I don't think it's a huge deal, still better than linker errors anyway. 


Differential Revision: [D26373027](https://our.internmc.facebook.com/intern/diff/D26373027)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/UpSampleNearest3d.cpp:
```diff
@@ -3,145 +3,79 @@
-      output_height,
-      output_width);
-
-  output.resize_({nbatch, channels, output_depth, output_height, output_width}, input.suggest_memory_format());
```
@fmassa @bdhirsh This looks like a bug; previously `resize_` took the suggested memory format from input, but the call to `suggest_memory_format` is completely lost in the code below.

fmassa replied:
cc @vfdev-5 can you try changing this locally and see if it fixes the behavior you were seeing?

ezyang replied:
It looks like this bug also applies to upsample_nearest2d

ezyang replied:
(but not upsample_nearest1d, which seems to have never gotten memory format support)

fmassa replied:
Looks like we might need to audit the other places as well as for 2d this also seemed to happen https://github.com/pytorch/pytorch/pull/52012/files#r589563928

bdhirsh replied:
Sorry about that :/ Happy to go through the PR's in this stack and fix that if someone hasn't started already

fmassa replied:
@bdhirsh if you could fix it it would be great

ngimel replied:
Additional wrinkle here is that cuda seems to not support layout propagation for this op, and backward layout propagation is not supported neither for cuda nor cpu (unless I'm overlooking something).

bdhirsh replied:
leaving the fix here for the paper trail: https://github.com/pytorch/pytorch/pull/53535/
